envoy.dependency.cve_scan/tests/test_integration.py

```diff
@@ -0,0 +1,194 @@
+
+import gzip
+import json
+import pathlib
+from functools import cached_property
+from unittest.mock import AsyncMock, PropertyMock
+from typing import Type
+
+import yaml
+
+import pytest
+
+import abstracts
+
+from envoy.dependency import cve_scan
+
+
+#
+# As `envoy.dependency.cve_scan` only provides an abstract
+# implementation, all classes need to be overridden and wired together.
+#
+# The advantage of this is that implementations have all of the
+# required classes available for customisation or debugging, and
+# the structure of the package is explicit.
+#
+# The integration tests here provide a reference implementation
+#
+
+#
+# *CVEChecker implementation*
+#
+
+@abstracts.implementer(cve_scan.ACVE)
+class DummyCVE:
+
+    @property
+    def cpe_class(self):
+        return DummyCPE
+
+    @property
+    def version_matcher_class(self) -> Type[cve_scan.ACVEVersionMatcher]:
+        return DummyCVEVersionMatcher
+
+
+@abstracts.implementer(cve_scan.ACPE)
+class DummyCPE:
+    pass
+
+
+@abstracts.implementer(cve_scan.ADependency)
+class DummyDependency:
+    pass
+
+
+@abstracts.implementer(cve_scan.ACVEChecker)
+class DummyCVEChecker:
+
+    @property
+    def cpe_class(self):
+        return DummyCPE
+
+    @property
+    def cve_class(self):
+        return DummyCVE
+
+    @property
+    def dependency_class(self):
+        return DummyDependency
+
+    # implementations should cache this
+    @property
+    def dependency_metadata(self):
+        # this needs to be mocked for these tests
+        return super().dependency_metadata
+
+    @cached_property
+    def ignored_cves(self):
+        return super().ignored_cves
+
+
+@abstracts.implementer(cve_scan.ACVEVersionMatcher)
+class DummyCVEVersionMatcher:
+    pass
+
+
+#
+# *Test data*
+#
+
+configs = pathlib.Path("envoy.dependency.cve_scan/tests/integration/config")
+CONFIG_TEST_DATA = {
+    f"config:{p.name.split('.')[0]}":
+    p for p in list(configs.glob("*.yaml"))}
+nist = pathlib.Path("envoy.dependency.cve_scan/tests/integration/nist")
+NIST_TEST_DATA = {
+    f"nist:{p.name.split('.')[0]}":
+    p for p in list(nist.glob("*.json"))}
+deps = pathlib.Path("envoy.dependency.cve_scan/tests/integration/deps")
+DEPS_TEST_DATA = {
+    f"deps:{p.name.split('.')[0]}": p
+    for p in list(deps.glob("*.json"))}
+outcomes = pathlib.Path("envoy.dependency.cve_scan/tests/integration/outcomes")
+OUTCOME_TEST_DATA = {
+    f"outcome:{p.name.split('.')[0]}": p
+    for p in list(outcomes.glob("*.json"))}
```

Comment:
Is there a way to refactor this to avoid boilerplate?

Reply:
iiuc the suggestion probably yes - i can just make them all into a single `TEST_DATA` dictionary and populate dynamically...